### PR TITLE
[PowerPC] Diagnose invalid combination with Altivec, VSX and soft-float

### LINKED ISSUE #55556: PPC Altivec: Killed - processing time exceeded
**Testcase**:

```c++
extern void flag(unsigned long);

static __inline__ int
vec_all_eq(vector unsigned char __a, vector unsigned char __b) {
  return __builtin_altivec_vcmpequb_p(2, (vector char)__a, (vector char)__b);
}

static __inline__ vector unsigned char
    vec_perm(vector unsigned char __a, vector unsigned char __b,
             vector unsigned char __c) {

  return (vector unsigned char)__builtin_altivec_vperm_4si(
      (vector int)__a, (vector int)__b, __c);
}

int t2(void) {
  unsigned long msr;

  __vector unsigned char a = {0xA0, 0xA1, 0xA2, 0xA3, 0xA4, 0xA5, 0xA6, 0xA7,
                               0xA8, 0xA9, 0xAA, 0xAB, 0xAC, 0xAD, 0xAE, 0xAF};

  __vector unsigned char b = {0xB0, 0xB1, 0xB2, 0xB3, 0xB4, 0xB5, 0xB6, 0xB7,
                               0xB8, 0xB9, 0xBA, 0xBB, 0xBC, 0xBD, 0xBE, 0xBF};

  __vector unsigned char c = {0x01, 0x14, 0x18, 0x10, 0x06, 0x15,
                                  0x19, 0x1A, 0x1C, 0x1C, 0x1C, 0x13,
                                  0x08, 0x1D, 0x1B, 0x0E};

  __vector unsigned char s = {161, 180, 184, 176, 166, 181, 185, 186,
                                 188, 188, 188, 179, 168, 189, 187, 174};

  __vector unsigned char r;

  r = vec_perm(a, b, c);

  if (!vec_all_eq(s, r)) {
    flag(msr);
    return 1;
  }

  return 0;
}
```

**Compile options**:
```shell
-mcpu=a2 -msoft-float --target=ppc32 -O2 -maltivec
```

**Error**:
```console
Killed - processing time exceeded
Compiler returned: 143
```

### PR DESCRIPTION
**Pull request by chenzheng1030**
Moved from https://reviews.llvm.org/D126302

The current behaviour with these three options is quite undesirable:
-mno-altivec -mvsx allows VSX to override no Altivec, thereby turning on both
-msoft-float -maltivec causes a crash if an actual Altivec instruction is required because soft float turns of Altivec
-msoft-float -mvsx is also accepted with both Altivec and VSX turned off (potentially causing crashes as above)

This patch diagnoses these impossible combinations in the driver so the user does not end up with surprises in terms of their options being ignored or silently overridden.

Fixes https://github.com/llvm/llvm-project/issues/55556

### CODE REVIEW

File: clang/test/CodeGen/PowerPC/attr-target-ppc.c [hunk 1]
```diff
@@ -1,4 +1,7 @@
 // RUN: not %clang_cc1 -triple powerpc64le-linux-gnu -emit-llvm %s -o -
 
 long __attribute__((target("power8-vector,no-vsx"))) foo (void) { return 0; }  // expected-error {{option '-mpower8-vector' cannot be specified with '-mno-vsx'}}
+long __attribute__((target("no-altivec,vsx"))) foo2(void) { return 0; }        // expected-error {{option '-mvsx' cannot be specified with '-mno-altivec'}}
+long __attribute__((target("no-hard-float,altivec"))) foo3(void) { return 0; } // expected-error {{option '-msoft-float' cannot be specified with '-maltivec'}}
+long __attribute__((target("no-hard-float,vsx"))) foo3(void) { return 0; } // expected-error {{option '-msoft-float' cannot be specified with '-mvsx'}}
```
Comment (amy-kwan) [inline clang/test/CodeGen/PowerPC/attr-target-ppc.c:6]: nit: Rename as `foo4`?

**Comment by chenzheng1030** [comment]
Patch updated.

**Review by amy-kwan — APPROVED** [review]
I don't have any further comments, so I think LGTM.

**Comment by chenzheng1030** [comment]
> I don't have any further comments, so I think LGTM.

Thanks Amy.

**Comment by brad0** [comment]
cc @MaskRay

**Comment by MaskRay** [comment]
> cc @MaskRay

The change is reasonable. I'd indent continuation lines by 2 spaces, but probably not worth the churn.

**Pull request merged**